File: ml/udm/04_polynomial_regression/data_preprocessing_template.R
# Data Preprocessing Template

# Importing the dataset
dataset = read.csv('Position_Salaries.csv')
dataset = dataset[2:3]

# Splitting the dataset into the Training set and Test set
# install.packages('caTools')
# library(caTools)
# set.seed(123)
# split = sample.split(dataset$DependentVariable, SplitRatio = 0.8)
# training_set = subset(dataset, split == TRUE)
# test_set = subset(dataset, split == FALSE)

# Feature Scaling
# training_set = scale(training_set)
# test_set = scale(test_set)

# fitting linear regression to dataser

lin_reg = lm(
  formula = Salary ~ Level,
  data = dataset
)

# fitting polynomial regression to dataser
dataset$Level2 = dataset$Level^2
dataset$Level3 = dataset$Level^3
dataset$Level4 = dataset$Level^4
lin_reg2 = lm(
  formula = Salary ~.,
  data = dataset
)

# visualizing 
library(ggplot2)
ggplot() +
  geom_point(
    aes(
      x = dataset$Level,
      y = dataset$Salary
    ),
    colour='red'
  ) +
  geom_line(
    aes(
      x = dataset$Level,
      y = predict(lin_reg, newdata = dataset)
    ),
    colour='blue'
  ) +
  geom_line(
    aes(
      x = dataset$Level,
      y = predict(lin_reg2, newdata = dataset)
    ),
    colour='green'
  )

# predicting
y_pred1 = predict(lin_reg, newdata = data.frame(Level = 6.5))
y_pred2 = predict(lin_reg2, newdata = data.frame(Level = 6.5, 
                                                 Level2 = 6.5^2, 
                                                 Level3 = 6.5^3, 
                                                 Level4 = 6.5^4))

File: ml/udm/04_polynomial_regression/data_preprocessing_template.py
# Data Preprocessing Template

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Position_Salaries.csv')
# make X a matrix
X = dataset.iloc[:, 1:2].values
y = dataset.iloc[:, 2].values

# Splitting the dataset into the Training set and Test set
# from sklearn.cross_validation import train_test_split
# X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

# Feature Scaling
"""from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)
sc_y = StandardScaler()
y_train = sc_y.fit_transform(y_train)"""

# fitting linear regression to the dataset
from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg.fit(X, y)

# fitting polynomial regression to the dataset
from sklearn.preprocessing import PolynomialFeatures
poly_reg = PolynomialFeatures(degree=4)
X_poly = poly_reg.fit_transform(X)
lin_reg2 = LinearRegression()
lin_reg2.fit(X_poly, y)

# visualizing linear and polynomial regression results
X_grid = np.arange(min(X), max(X), 0.01)
X_grid = X_grid.reshape((len(X_grid), 1))
plt.scatter(X, y, color='red')
plt.plot(X, lin_reg.predict(X), color='blue')
plt.plot(X_grid, lin_reg2.predict(poly_reg.fit_transform(X_grid)), color='green')
plt.title('Salary vs. Level')
plt.xlabel('Level')
plt.ylabel('Salary')
plt.show()

# predicting with linear regression
lin_reg.predict(6.5)
y_lin_pred = lin_reg.predict(X)
# predixting with polymonial regression
lin_reg2.predict(poly_reg.fit_transform(6.5))
y_poly_pred = lin_reg2.predict(poly_reg.fit_transform(X))
